### PR TITLE
feat: automate updates of api & sdk docs

### PR DESCRIPTION
**Pull request by chadoh**
**Question**: Given that the API docs for ~nearlib~ [`near-api-js`](https://github.com/near/near-api-js) are [auto-generated](https://github.com/near/near-api-js/blob/8df54f254caefbdddcbd032c79d928c8991bc756/package.json#L62), does it make sense for a user to end up at these auto-generated docs when they click the "Edit" link while browsing our docs? (This same question applies to [`near-sdk-as`](https://github.com/near/near-sdk-as) and [`near-sdk-rs`](https://github.com/near/near-sdk-rs))

**Answer**: No. We wouldn't want someone to contribute edits to these auto-generated docs, as their changes would be lost the next time the docs are updated from the source.

So here's what this does
------------------------

When you run `yarn build` in this repository:

1. Shallow clone (or update) each source repository to a location that doesn't get checked into the repo
2. Run the command to build docs for each of these source repositories
3. Copy generated docs to main `docs` folder
4. Run the build step we've been using
5. Clean up – remove the auto-generated stuff from the `docs` folder

Note that this also allows us to stop checking these auto-generated docs files into git. Both here and in the source repositories. Less noise all around! 🎉 

This PR also improves the sidebar:

image

Caveat
------

Making the "Edit" link on each page link to _the relevant file_ on GitHub requires building a custom Markdown generator to use in each source project. For now, this PR makes the Edit link point to the root repository for each set of auto-generated docs, except for `near-sdk-rs`. See new [add-custom-edit-url](https://github.com/nearprotocol/docs/pull/291/files#diff-d8a990cb3771a02d01a27115c1b8eda2) script for details.

To do
-----

- [x] in build script, fetch directory from source repositories
- [ ] set up deploy hook so pushes to source repos kick off a rebuild & deploy this website
   - `near/near-api-js`
   - `near/near-sdk-as`
   - `near/near-sdk-rs`
- [x] in render settings, redirect from nearlib to new near-api-js endpoint _unimportant; the previous pages/URLs were probably not in wide use_
- [x] update "Edit" button to link to near-api-js source

### CODE REVIEW

**Comment by amgando** [comment]
build is breaking bc of `nearlib` links.  once we merge the renaming work asap in #295 then we can rebase off master and merge this.  does that make sense?

File: website/sidebars.json [hunk 1]
```diff
@@ -61,7 +61,7 @@
         "type": "subcategory",
         "label": "Creating Contracts",
         "ids": [
-          "near-bindgen/near-bindgen",
+          "api/near-sdk-rs/index",
```
Comment (chadoh) [inline website/sidebars.json:64]: @amgando note that this link doesn't show up in the sidebar: https://docs-pr-291.onrender.com/docs/roles/developer/contracts/assemblyscript

But the same exact `id` used on the `api` page, down on [line 161](https://github.com/nearprotocol/docs/pull/291/files#diff-cb1f944b76e14c5b994bf7231b98104aR161) below, _does_ generate the expected result in the sidebar: https://docs-pr-291.onrender.com/docs/api/near-sdk-rs/

Any idea how to fix this?

Comment (chadoh) [reply website/sidebars.json:64]: Hmm, it looks like we might need to update to Docusaurus v2 to fix this? https://github.com/facebook/docusaurus/issues/868

How do you think we should proceed in the meantime?

**Comment by mikedotexe** [comment]
I did:
`cd website`
`yarn && yarn build`
got
```bash
/Users/mike/near/docs/website/bin/replace-snippets.js:17
  from: replacements["replacements"][i]["from"],
        ^

ReferenceError: replacements is not defined
…
error Command failed with exit code 1.
```

File: website/bin/replace-snippets.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+const options = {
+
+  // Glob(s)
+  files: [
+    path.join(__dirname, '/../tests/*'),
+    path.join(__dirname, '/../../docs/**/*')
+  ],
+
+  // Replacement to make (string or regex)
+  from: replacements["replacements"][i]["from"],
```
Comment (mikedotexe) [inline website/bin/replace-snippets.js:17]: yeah think we need to define replacements first, maybe you're about to commit this, just a reminder

Comment (chadoh) [reply website/bin/replace-snippets.js]: Updated!

File: docs/tutorials/how-to-write-contracts-that-talk-to-each-other.md [hunk 3]
```diff
@@ -354,7 +354,7 @@ export class CalculatorApi {
 }
 ```
 
-*(For more info on `ContractPromise`, check out [ContractPromise](/docs/runtime-ts/classes/contractpromise) and [ContractPromiseResult](/docs/runtime-ts/classes/contractpromiseresult))*
+*(For more info on `ContractPromise`, check out [ContractPromise](/docs/api/near-sdk-as/classes/_runtime_contract_.contractpromise) and [ContractPromiseResult](/docs/api/near-sdk-as/classes/_runtime_contract_.contractpromiseresult))*
```
Comment (chadoh) [inline docs/tutorials/how-to-write-contracts-that-talk-to-each-other.md:357]: I'm not sure what to link to here

File: docs/tutorials/zero-to-hero.md [hunk 4]
```diff
@@ -62,7 +62,7 @@ Now that you're familiar with the various files in your project, let's get start
 
 The first thing that our Oracle Contract must be able to do is read from and write to the blockchain. This way our contract can save external data onto the blockchain for other contracts to interact with.
 
-Navigate to our API [storage docs](/docs/runtime-ts/classes/storage) to review the `setString` and `getString` functions. In later steps we'll show you how to handle more complicated data types.
+Navigate to our API [storage docs](/docs/api/near-sdk-as/classes/_runtime_storage_.storage) to review the `setString` and `getString` functions. In later steps we'll show you how to handle more complicated data types.
```
Comment (chadoh) [inline docs/tutorials/zero-to-hero.md:65]: I'm not sure what to link to here, either

File: website/sidebars.json (hunk 1 above)
Comment (amgando) [reply website/sidebars.json:64]: only the first `id` wins I think, or the last, can't recall. we don't need the page in the other location. let's just put all API docs in this section

I don't think we should rush docusaurus v2. after RL1 maybe? wdyt?

File: docs/tutorials/how-to-write-contracts-that-talk-to-each-other.md (hunk 3 above)
Comment (amgando) [reply docs/tutorials/how-to-write-contracts-that-talk-to-each-other.md:357]: `runtime-ts` is now `near-sdk-as`. these links would go to equivalents in the new auto generated docs folders

File: docs/tutorials/zero-to-hero.md (hunk 4 above)
Comment (amgando) [reply docs/tutorials/zero-to-hero.md:65]: same as above

**Comment by chadoh** [comment]
After discussing this with @amgando, we have decided on a much simpler approach to this problem

**Pull request closed without merging**